crates/ruma-state-res/src/event_auth.rs

```diff
@@ -731,23 +740,44 @@ fn check_power_levels(
         }
     }
 
+    // - If any of the keys users_default, events_default, state_default, ban, redact, kick, or
+    //   invite in content are present and not an integer, reject.
+    // - If either of the keys events or notifications in content are present and not a dictionary
+    //   with values that are integers, reject.
+    // - If users key in content is not a dictionary with keys that are valid user IDs with values
+    //   that are integers, reject.
+    let user_content: RoomPowerLevelsEventContent = if room_version.integer_power_levels {
+        match from_json_str::<IntRoomPowerLevelsEventContent>(power_event.content().get()) {
+            Ok(content) => content.into(),
+            Err(_) => {
+                error!("m.room.power_levels event is not valid with integer values");
+                return None;
+            }
+        }
+    } else {
+        match from_json_str(power_event.content().get()) {
+            Ok(content) => content,
+            Err(_) => {
+                error!(
+                    "m.room.power_levels event is not valid with integer or string integer values"
+                );
+                return None;
+            }
+        }
+    };
+
+    // Validation of users is done in Ruma, synapse for loops validating user_ids and integers here
+    info!("validation of power event finished");
+
     let current_state = match previous_power_event {
         Some(current_state) => current_state,
         // If there is no previous m.room.power_levels event in the room, allow
         None => return Some(true),
     };
```

Comment:
It should always have been done after validation (see the [auth rules](https://spec.matrix.org/unstable/rooms/v10/#authorization-rules)).

We cannot accept a power levels event that is malformed just because there was no previous event.